pkg/metrics/metrics.go

```diff
@@ -0,0 +1,276 @@
+package metrics
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/kanisterio/kanister/pkg/log"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// BoundedLabel is a type that represents a label and its associated
+// valid values
+type BoundedLabel struct {
+	LabelName   string
+	LabelValues []string
+}
+
+// getLabelNames extracts the all the `LabelName` fields from each BoundedLabel struct
+func getLabelNames(bl []BoundedLabel) []string {
+	ln := make([]string, 0)
+	for _, l := range bl {
+		ln = append(ln, l.LabelName)
+	}
+	return ln
+}
+
+// fillLabelCombinations uses a backtracking approach to generate
```

Comment:
While backtracking is a well-known CS technique, it is complicated and not easy to check for errors.

In the interest of code maintainability, I suggest either:
* Add a citation in comment to other open source or public article that explains this use.
* Or consider making use of an existing package such as https://pkg.go.dev/gonum.org/v1/gonum/stat/combin#Cartesian. If we pass a slice containing the lengths of each of the BoundedValue.LabelValue slices, the Cartesian product that is returned will contain an array of slices of integers and these can be used as indices into each of the LabelValue slices.

It appears to me that this function works and we have test cases, but for production code in industry I'd favor something simpler.

Reply:
Feel free to ask @viveksinghggits or @pavannd1 on the choice of algorithms. I'm not saying no, but raising a concern.

Reply:
Sure thank you. I'll explore these options and update this discussion thread with my findings.

Reply:
@ewhamilton Thanks! Using combinations.Cartesian did make things very simple - the code is now iterative instead of being a recursive implementation - we're also maintaining the same set of unit tests which are still passing!